bpf-gpl/tc.c

```diff
@@ -178,6 +178,20 @@ static CALI_BPF_INLINE int forward_or_drop(struct __sk_buff *skb,
 		if  (CALI_F_FROM_HOST) {
 			redir_flags = BPF_F_INGRESS;
 		}
+
+		/* Revalidate the access to the packet */
+		if ((void *)(long)skb->data + sizeof(struct ethhdr) > (void *)(long)skb->data_end) {
+			reason = CALI_REASON_SHORT;
+			goto deny;
+		}
+
+		/* Swap the MACs as we are turning it back */
+		struct ethhdr *eth_hdr = (void *)(long)skb->data;
+		unsigned char mac[ETH_ALEN];
+		__builtin_memcpy(mac, &eth_hdr->h_source, ETH_ALEN);
```

Comment:
This needs to be from `h_dest`, not `h_source`.

(DST MAC -> buffer, SOURCE MAC -> DST MAC, buffer -> SOURCE MAC)

Reply:
yes